lib/tramway/base_form.rb

```diff
@@ -30,9 +30,22 @@ def properties(*attributes)
             property(attribute)
           end
         else
-          @properties || []
+          __properties
         end
       end
+
+      def __properties
+        @properties ||= []
```

Comment:
I see that you're using `@properties ||= []` several times in the base_form.rb class. You could initializes the @properties class instance variable one time when the class is loaded

```
class << self
   @properties = []

  def property(attribute)
    (__ancestor_properties + @properties).uniq
  end
end
```

Reply:
I've never seen this approach. Are you sure that initializing class-level variables out of methods is nice? :slightly_smiling_face:

Reply:
It's for sure nicer to have one place of default initialization, and I dont see problems with class level variable.

But I also haven't used this approach much, so can't comment if it have no problems at all

Reply:
I've tested this approach. For some reason, `@properties` is not initialized on class load.

Reply:
Ah, I see, the variable initialization doesn't pass to child classes. Sorry, then, my bad :) 
The initialization still could be isolated with `inherited` hook, if you willing to go that path, like this

```ruby
class Parent
  def self.inherited(subclass)
    subclass.instance_variable_set(:@things, [])
  end

  class << self
    def add(value)
      @things << value
    end
  end
end

class ChildDecorator < Parent
  add :third
  add :four
end
```

Reply:
Nice! :slightly_smiling_face: Thanks!!

btw, Linters are v ahuie :laughing:

Reply:
> btw, Linters are v ahuie 

They are right, it should be `@properties`, not `@@properties` :) @@properties will give you effects you don't want, like - if you change child properties, the parent properties also will be changed

Reply:
Maybe I missed something, but looks like we can't use `@properties` in our case.

```ruby
subclass.class_variable_set(:@properties, __ancestor_properties)
```

return `NameError: '@properties' is not allowed as a class variable name`

Reply:
because it should be `instance_variable_set`, not `class_variable_set`. As the class is actually instance of Class itself